code/game/mecha/mech_prosthetics.dm

```diff
@@ -0,0 +1,365 @@
+		progress = 0
+	queue.Cut(index, index + 1)
+	update_busy()
+
+/obj/machinery/pros_fabricator/proc/can_build(var/datum/design/D)
+	for(var/M in D.materials)
+		if(materials[M] < D.materials[M])
+			return 0
+	return 1
+
+/obj/machinery/pros_fabricator/proc/check_build()
+	if(!queue.len)
+		progress = 0
+		return
+	var/datum/design/D = queue[1]
+	if(!can_build(D))
+		progress = 0
+		return
+	if(D.time > progress)
+		return
+	for(var/M in D.materials)
+		materials[M] = max(0, materials[M] - D.materials[M] * mat_efficiency)
+	if(D.build_path)
+		var/obj/new_item = D.Fabricate(get_step(get_turf(src), src.dir), src)
+		visible_message("\The [src] pings, indicating that \the [D] is complete.", "You hear a ping.")
+		if(mat_efficiency != 1)
+			if(new_item.matter && new_item.matter.len > 0)
+				for(var/i in new_item.matter)
+					new_item.matter[i] = new_item.matter[i] * mat_efficiency
+	remove_from_queue(1)
+
+/obj/machinery/pros_fabricator/proc/get_queue_names()
+	. = list()
+	for(var/i = 2 to queue.len)
+		var/datum/design/D = queue[i]
+		. += D.name
+
+/obj/machinery/pros_fabricator/proc/get_build_options()
+	. = list()
+	for(var/i = 1 to files.known_designs.len)
+		var/datum/design/D = files.known_designs[i]
+		if(D.build_path && (D.build_type & PROSFAB))
+			. += list(list("name" = D.name, "id" = i, "category" = D.category, "resourses" = get_design_resourses(D), "time" = get_design_time(D)))
+
+/obj/machinery/pros_fabricator/proc/get_design_resourses(var/datum/design/D)
+	var/list/F = list()
+	for(var/T in D.materials)
+		F += "[capitalize(T)]: [D.materials[T] * mat_efficiency]"
+	return english_list(F, and_text = ", ")
+
+/obj/machinery/pros_fabricator/proc/get_design_time(var/datum/design/D)
+	return time2text(round(10 * D.time / speed), "mm:ss")
+
+/obj/machinery/pros_fabricator/proc/update_categories()
+	categories = list()
+	for(var/datum/design/D in files.known_designs)
+		if(!D.build_path || !(D.build_type & PROSFAB))
+			continue
+		categories |= D.category
+	if(!category || !(category in categories))
+		category = categories[1]
+
+/obj/machinery/pros_fabricator/proc/get_materials()
+	. = list()
+	for(var/T in materials)
+		. += list(list("mat" = capitalize(T), "amt" = materials[T]))
+
+/obj/machinery/pros_fabricator/proc/eject_materials(var/material, var/amount) // 0 amount = 0 means ejecting a full stack; -1 means eject everything
+	var/recursive = amount == -1 ? 1 : 0
+	material = lowertext(material)
+	var/mattype
+	switch(material)
+		if(DEFAULT_WALL_MATERIAL)
+			mattype = /obj/item/stack/material/steel
+		if("glass")
+			mattype = /obj/item/stack/material/glass
+		if("gold")
+			mattype = /obj/item/stack/material/gold
+		if("silver")
+			mattype = /obj/item/stack/material/silver
+		if("diamond")
+			mattype = /obj/item/stack/material/diamond
+		if("phoron")
+			mattype = /obj/item/stack/material/phoron
+		if("uranium")
+			mattype = /obj/item/stack/material/uranium
+		if("plasteel")
+			mattype = /obj/item/stack/material/plasteel
```

Comment:
https://github.com/PolarisSS13/Polaris/blob/master/code/modules/materials/materials.dm

Stacks have a `material` var holding their material; if you *need* a string, `material.name` is the material names you have here. `get_material_by_name(name)` will get you the material datum from the string; you'll want `[material].place_sheet(sheet_loc)`.